### PR TITLE
fix : Throw an error if graphiql uses a restricted route

### PR DESCRIPTION
**Pull request by Vishal-Sikka**
whether the enableGraphiql route is a restricted route it will throw error

fix: #349 
/claim #349

### CODE REVIEW

File: src/http/server.rs [hunk 1]
```diff
@@ -56,20 +56,31 @@ fn create_allowed_headers(headers: &HeaderMap, allowed: &HashSet<String>) -> Hea
...
+    
+    let make_svc = make_service_fn(move |_conn| {
+        let state = Arc::clone(&state);
+        async move { Ok::<_, anyhow::Error>(service_fn(move |req| handle_request(req, state.clone()))) }
+    });
+    
+    let addr = ([0, 0, 0, 0], port).into();
+    let server = hyper::Server::try_bind(&addr).map_err(CLIError::from)?.serve(make_svc);
+    
+    Ok(server.await.map_err(CLIError::from)?)
```
Comment (tusharmath) [inline src/http/server.rs:85]: Need to write an integration test. Check the errors folder inside tests.

Comment (Vishal-Sikka) [reply src/http/server.rs:85]: written test

File: tests/graphql/errors/test-restricted-graphiql-route.graphql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#> server-sdl
+schema {
+    query: Query
```
Comment (tusharmath) [inline tests/graphql/errors/test-restricted-graphiql-route.graphql:3]:
```suggestion
    query: Query
    enableGraphiql: "/graphql"
```

This is how the test should be implemented.

File: src/http/server.rs (hunk 1 above)
Comment (tusharmath) [reply src/http/server.rs:85]: The assertion should happen inside of `from_config.rs` not in `server.rs`

File: tests/graphql/errors/test-restricted-graphiql-route.graphql [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    enableGraphiql: "/graphql"
+}
+
+type Query {
+    hello: String
+}
+
+#> client-sdl
+type Failure
+@error(message: "Cannot use restricted routes '/graphql' for enabling graphiql", trace: [])
```
Comment (tusharmath) [inline tests/graphql/errors/test-restricted-graphiql-route.graphql:13]:
```suggestion
@error(message: "Cannot use restricted route '/graphql' for graphiql", trace: ["schema", "@server"])
```

File: src/blueprint/from_config.rs [hunk 4]
```diff
@@ -24,13 +25,34 @@ use crate::{blueprint, config};
...
+  Ok(())
+}
+
 pub fn config_blueprint(config: &Config) -> Valid<Blueprint> {
+  validate_route(config)?;
   let output_types = config.output_types();
   let input_types = config.input_types();
   let schema = to_schema(config)?;
   let definitions = to_definitions(config, output_types, input_types)?;
   Ok(super::compress::compress(Blueprint { schema, definitions }))
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:53]:
```suggestion
  validate_route(config).validate_or({
    let output_types = config.output_types();
    let input_types = config.input_types();
    let schema = to_schema(config)?;
    let definitions = to_definitions(config, output_types, input_types)?;
    Ok(super::compress::compress(Blueprint { schema, definitions }))
  })
```

File: src/blueprint/from_config.rs [hunk 5]
```diff
@@ -24,13 +25,34 @@ use crate::{blueprint, config};
...
+    ).trace(vec!["schema".into(), "server".into(), "enableGraphiql".into()]);
+
+    return Err(
+      ValidationError::new(
+        format!("Cannot use restricted routes '{}' for enabling graphiql", enable_graphiql)
+      )
+      );
+    }
+  }
+  Ok(())
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:44]: Can simplify this logic.

Comment (Vishal-Sikka) [reply src/blueprint/from_config.rs:44]: ok i will simply it more, we need to throw error with trace right , if  only validation error works then another idea is to do validation in config_blueprint function it self

Comment (tusharmath) [reply src/blueprint/from_config.rs:44]: You can perform a `Valid::fail` like we have done in other places.

File: tests/graphql/errors/test-restricted-graphiql-route.graphql [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    enableGraphiql: "/graphql"
+}
+
+type Query {
+    hello: String
+}
+
+#> client-sdl
+type Failure
+@error(message: "Cannot use restricted route '/graphql' for graphiql", trace: [schema, server, enableGraphiql])
```
Comment (tusharmath) [inline tests/graphql/errors/test-restricted-graphiql-route.graphql:13]:
```suggestion
@error(message: "Cannot use restricted route '/graphql' for IDE", trace: ["schema", "@server", "enableGraphiql"])
```

Let's update the implementation to match the above error

File: src/blueprint/from_config.rs [hunk 7]
```diff
@@ -24,13 +24,29 @@ use crate::{blueprint, config};
...
+fn validate_route(config: &Config) -> Valid<()> {
+    match config.server.enable_graphiql.as_deref() {
+        Some(route) if RESTRICTED_ROUTES.contains(&route) => {
+            let error_message =
+                format!("Cannot use restricted routes '{}' for enabling graphiql", route);
+            Valid::fail(error_message).trace("[schema, server, enableGraphiql]")
+        }
+        _ => Valid::Ok(()),
+    }
+}
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:38]: I think using `if let Some` will make this code a little more simpler.

**Review by tusharmath — CHANGES_REQUESTED** [review]
Please fix build failure.

**Comment by Vishal-Sikka** [comment]
> Please fix build failure.
Screenshot 2023-09-30 at 3 01 39 PM

somehow build is not failing for me - updated my branch - check once again

**Comment by tusharmath** [comment]
Run `./lint.sh`.

**Comment by Vishal-Sikka** [comment]
> ./lint.sh

did

**Comment by tusharmath** [comment]
then commit and push :)

**Comment by tusharmath** [comment]
@Vishal-Sikka this should be a quick fix, I think. Let me know if you are still on this.

**Comment by Vishal-Sikka** [comment]
> @Vishal-Sikka this should be a quick fix, I think. Let me know if you are still on this.

just pushed now  running ./lint.sh not changing much, let me know if it getting build , for me it getting success

File: tests/graphql/errors/test-restricted-graphiql-route.graphql [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    enableGraphiql: "/graphql"
+}
+
+type Query {
+    hello: String
+}
+
+#> client-sdl
+type Failure
+@error(message: "Cannot use restricted route '/graphql' for IDE", trace: ["schema", "@server", "enableGraphiql"])
```
Comment (tusharmath) [inline tests/graphql/errors/test-restricted-graphiql-route.graphql:13]: The implementation and the tests don't match.

**Comment by tusharmath** [comment]
fixed in #447

**Pull request closed without merging**